app/scripts/metamask-controller.js

```diff
@@ -1168,6 +1185,148 @@ export default class MetamaskController extends EventEmitter {
     }
   }
 
+  // eth_decryptMessage methods
+
+  /**
+  * Called when a dapp uses the eth_decryptMessage method.
+  *
+  * @param {Object} msgParams - The params of the message to sign & return to the Dapp.
+  * @param {Function} cb - The callback function called with the signature.
+  * Passed back to the requesting Dapp.
+  */
+  async newRequestDecryptMessage (msgParams, req) {
+    const promise = this.decryptMessageManager.addUnapprovedMessageAsync(msgParams, req)
+    this.sendUpdate()
+    this.opts.showUnconfirmedMessage()
+    return promise
+  }
+
+  /**
+  * Only decypt message and don't touch transaction state
+  *
+  * @param {Object} msgParams - The params of the message to decrypt & return to the Dapp.
+  * @returns {Promise<Object>} - A full state update.
+  */
+  decryptMessageInline (msgParams) {
+    log.info('MetaMaskController - decryptMessageInline')
+    const msgId = msgParams.metamaskId
+    const stripped = ethUtil.stripHexPrefix(msgParams.data)
+    const buff = Buffer.from(stripped, 'hex')
+    msgParams.data = JSON.parse(buff.toString('utf8'))
+    // decrypt the message inline
+    return this.keyringController.decryptMessage(msgParams).then((rawData) => {
+      const msg = this.decryptMessageManager.getMsg(msgId)
+      msg.rawData = rawData
+      this.decryptMessageManager._updateMsg(msg)
+      return this.getState()
+    })
+  }
+
+  /**
+  * Signifies a user's approval to decrypt a message in queue.
+  * Triggers decrypt, and the callback function from newUnsignedDecryptMessage.
+  *
+  * @param {Object} msgParams - The params of the message to decrypt & return to the Dapp.
+  * @returns {Promise<Object>} - A full state update.
+  */
+  decryptMessage (msgParams) {
+    log.info('MetaMaskController - decryptMessage')
+    const msgId = msgParams.metamaskId
+    // sets the status op the message to 'approved'
+    // and removes the metamaskId for decryption
+    try {
+      return this.decryptMessageManager.approveMessage(msgParams)
```

Comment:
Could you update this to use `async/await` instead? This is a good example of why traditional Promises can be error-prone - if an error occurs in this chain, it does get returned but it won't be caught by the `catch` block.